cEP-0024.md

```diff
@@ -0,0 +1,255 @@
+[GSoC project](https://summerofcode.withgoogle.com/projects/#5344316227059712).
+
+## Introduction
+
+[GitMate](https://gitmate.io/) is an automation tool for developers. While it
+works well, there are a number of missing features which would save the time of
+coala community.
+
+Many of the plugins in GitMate are general plugins, but coala needs more
+plugins which support coala development workflow. So, this project
+adds plugins to auto-review, auto-reject, auto-label PRs, assign, unassign
+issues, squash commits.
+
+## Changes in coala workflow
+
+### Use increasing difficulty level order to assign issues
+
+The current assign process in coala is to check team membership of a user,
+difficulty level of the issue and assign the issue to user if eligible. To
+implement the current assign process in GitMate, IGitt would need to support
+teams.
+
+Currently, IGitt doesn't support teams and there are no teams in GitLab.
+Considering [GitLab groups](https://docs.gitlab.com/ee/user/group/) and
+[GitHub teams](https://help.github.com/articles/about-teams/) as teams in
+IGitt is a possible solution for IGitt implementation. But, it requires extra
+tools to sync GitHub team and GitLab group memberships, which is a manual
+process and isn't automated yet.
+
+Also, team membership information is not publicly accessible in GitHub, so any
+system which needs to implement the current assign system would require
+organization authentication tokens.
+
+The best possible solution would be using an increasing difficulty level order,
+so that a user can be assigned to an issue only if the user has solved issues
+of lower difficulty level before requesting to get assigned to a higher
+difficulty level issue.
+
+#### Drawbacks
+
+If a user completed the lower difficulty level issues in GitHub and want to
+work on a higher difficulty level issue in GitLab, he/she has to repeat the
+newcomer process in GitLab. To resolve this issue, GitMate needs to find a way
+to link GitHub org and GitLab org, this would be easy when they have the same
+name, as in the case of `coala`. But, resolving this issue is out of scope for
+this project.
+
+If a user is worthy enough to work on the higher level difficulty issues
+directly, then that user needs to be manually assigned by the maintainers until
+that user completes at least one low difficulty issue. Even with current assign
+system, manual intervention by maintainers is needed to add a user to
+developers team.
+
+## New Features
+
+### Auto-assign issue when it is mentioned in any of PR commits
+
+When a PR is opened or updated, if any of PR's commit message referenced
+any issues with keywords `Fixes` or `Closes`, there is no
+CI failure in PR, then auto assign the author of the PR to the mentioned
+unassigned issues. If the author of the PR isn't eligible to get assigned to
+the mentioned issues, missing eligibility conditions will be specified as a
+comment, which isn't configurable.
```

Comment:
You havent mentioned labeling of ineligible difficulty, which I requested at https://github.com/coala/cEPs/pull/142#discussion_r201557412

PRs which are beyond the current difficulty level eligibility of the PR author need to be labelled.  If we dont manage these, newcomers will disregard the difficulty levels and will just send a PR anyway.

And labels must be configurable because different orgs use different labelling schemes.
As far as I have seen, GitMate plugins alway let the org admins choose the labels.

So just to be clear, this *must* have a configurable label.
Obviously 'author/non-member' would not be appropriate for a member of the org.

It should not be hard to add a configurable message as I requested, but the need to config the message is unknown yet as I cant see what the hard-coded message will be.  My guess is that a default message would not be suitable for coala, as we would want to link to http://coala.io/newcomer , etc.

Reply:
Giving a configurable message might lead to a messy comment by GitMate, so I am avoiding that support as of now. We can look into it later.

Reply:
Please add here the hard-coded message that the plugin will use for this scenario.